lsp/src/main/kotlin/org/gradle/declarative/lsp/valueFactoryIndex.kt

```diff
@@ -0,0 +1,113 @@
+
+    private fun indexValueFactories(schema: AnalysisSchema): Map<FqName, List<ValueFactoryIndexEntry>> {
+        val factoryIndex = mutableMapOf<FqName, List<ValueFactoryIndexEntry>>()
+        factoryIndex.merge(indexValueFactoriesFromExternalFunctions(schema))
+        factoryIndex.merge(indexValueFactoriesFromTopLevelReceiver(schema))
+        return factoryIndex
+    }
+
+    private fun indexValueFactoriesFromExternalFunctions(schema: AnalysisSchema): Map<FqName, List<ValueFactoryIndexEntry>> {
+        val factoryIndex = mutableMapOf<FqName, List<ValueFactoryIndexEntry>>()
+        schema.externalFunctionsByFqName
+            .map { it.value }
+            .filter { isValueFactoryFunction(it) }
+            .forEach {
+                val indexKey = returnTypeOfValueFactoryFunction(it)
+                val valueFactoryIndexEntry = ValueFactoryIndexEntry(it, "")
+                factoryIndex.merge(indexKey, listOf(valueFactoryIndexEntry)) { oldVal, newVal -> oldVal + newVal }
+            }
+        return factoryIndex
+    }
+
+    private fun indexValueFactoriesFromTopLevelReceiver(schema: AnalysisSchema) =
+        indexValueFactories(schema, schema.topLevelReceiverType, "")
+
+    private fun indexValueFactories(
+        schema: AnalysisSchema,
+        type: DataClass,
+        namePrefix: String
+    ): Map<FqName, List<ValueFactoryIndexEntry>> {
+        val factoryIndex = mutableMapOf<FqName, List<ValueFactoryIndexEntry>>()
+
+        type.memberFunctions
+            .filter { isValueFactoryFunction(it) }
+            .forEach {
+                val indexKey = returnTypeOfValueFactoryFunction(it)
+                val valueFactoryIndexEntry = ValueFactoryIndexEntry(it, namePrefix)
+                factoryIndex.merge(indexKey, listOf(valueFactoryIndexEntry)) { oldVal, newVal -> oldVal + newVal }
+            }
+
+        val typeRefContext = SchemaTypeRefContext(schema)
+        type.properties.forEach {
+            when (val propType = typeRefContext.resolveRef(it.valueType)) {
+                is DataClass -> {
+                    val propName = it.name
+                    val propIndex = indexValueFactories(schema, propType, "$namePrefix${propName}.")
+                    factoryIndex.merge(propIndex)
+                }
+
+                else -> Unit
+            }
+        }
+
+        return factoryIndex
+    }
+
+    private fun isValueFactoryFunction(function: SchemaFunction) =
+        function.semantics is FunctionSemantics.Pure && (function.returnValueType is DataTypeRef.Name || function.returnValueType is DataTypeRef.NameWithArgs)
+
+    private fun returnTypeOfValueFactoryFunction(it: SchemaFunction) =
+        when (val returnType = it.returnValueType) {
+            is DataTypeRef.Name -> returnType.fqName
+            is DataTypeRef.NameWithArgs -> returnType.fqName
+            is DataTypeRef.Type -> error("unexpected return type")
```

Comment:
Value factories of `Int`, `Boolean`, or `String` will have `DataTypeRef.Type` as their return type reference.

Reply:
Ok, but this `returnTypeOfValueFactoryFunction(SchemaFunction)` function is used in a way that it should never get a parameter whose `returnValueType` is a `DataTypeRef.Type`. See `isValueFactoryFunction(SchemaFunction)`.